### ISSUE TITLE
Random kernel cache dir for MPI programs

### ISSUE THREAD

**Issue by noma** [pocl/pocl#430]
I'm using PoCL within an MPI program, i.e. I have multiple communicating processes of the same program. Each of these processes (or ranks) uses PoCL to compute stuff. This scenario seems to collide with the kernel cache mechanism, as multiple processes seem to try to use the same cache directory, which leads to different kinds of non-deterministic program crashes.

It works if I use at most one process per compute node and a node-local filesystem for the cache to fully avoid conflicts. However, for performance reasons, I need more than one process per node. So I was wondering if there already is some intended way handle this scenario. One work-around I can think of would be a wrapper script around the application which gives every process an individual `POCL_CACHE_DIR` value.

**Comment by pjaaskel** [comment]
Process ID in the cache dir defeats the purpose of the cache, doesn't it? Does it work if you disable cache? The proper solution would be to have file locks in the cache updates.

**Comment by noma** [comment]
As a user, I don't really care about the cache. I noticed PoCL has long compilation times compared with vendor-implementations, but compared to the job runtimes, its negligible. 

The problem seems to be that POCL needs files to interact with LLVM. So setting `POCL_KERNEL_CACHE=0` does not help. According to the documentation, this only triggers deleting cache files with `clReleaseProgram()`. So I guess as long as this cannot be prevented, different cache paths per process are mandatory to avoid conflicts between processes. Otherwise I see errors ranging from segfaults in some processes, to failed asserts, related to file locking.

I currently wrap my application with a small bash script:
```bash
#!/bin/bash
EXE=my_executable
export POCL_CACHE_DIR=$(mktemp -d)
$EXE $@
```

In an ideal world, a cache with file locking etc. would lead to building everything only once across all processes of the MPI job. But I am not sure how safe/stable the whole caching is when it comes to compiling the same source with different compiler options/defines on the different processes. I remember having had some trouble which seemed like the caching mechanism used outdated compilation results instead of rebuilding. But that was some time ago, and I was too lazy to trace it back, so I just deleted the cache as a work-around.

Long story short: it might be a good addition to PoCL to have some environment variable that triggers using unique temporary cache dirs for each process (in addition to disable caching). Or maybe even fix the caching to work in an MPI-like scenario with multiple processes sharing a cache on a local/global filesystem.

This is not super urgent or a show stopper, as I can work-around it. But it's not too straight forward to find the actual problem/solution if you run into this and aren't familiar with PoCL.

**Comment by pjaaskel** [comment]
Yep, this known issue is not nice (https://bugs.launchpad.net/pocl/+bug/1039565)  @kraiskil implemented a midway solution for this, but there's still the code gen phase that requires file system as it doesn't call LLVM APIs for assembly and linking. As this does not affect our group's activities at the moment, we cannot put time on this in the near future, but any contributions are greatly appreciated and I will certainly review any PRs thrown my way.

The caching mechanism, on the other hand, is very much used in our embedded work. If it doesn't take in account the build options in the cache hash, it clearly is a bug. I opened Issue #434 to remember to check this out.

**Comment by pjaaskel** [comment]
I noticed there's code similar to what you are after (per process cache dir) for Android in pocl_cache.c if you wish, you might want to check if that can be adapted to your needs as well.

**Comment by franz** [comment]
It is a bug in pocl's cache code, will be fixed soon.
